File: Euclid.Storage/Database/Access/ItemDao.cs
using Euclid.Storage.Database.Data;
using NHibernate.Linq;
using System.Collections.Generic;
using System.Linq;

namespace Euclid.Storage.Database.Access
{
    public class ItemDao
    {
        /// <summary>
        /// Get list of all definition data
        /// </summary>
        public static List<ItemDefinitionData> GetDefinitions()
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                return session.QueryOver<ItemDefinitionData>().List() as List<ItemDefinitionData>;
            }
        }

        /// <summary>
        /// Get list of all item data for user inventory
        /// </summary>
        public static List<ItemData> GetUserItems(int userId)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                return session.QueryOver<ItemData>().Where(x => x.OwnerId == userId && x.RoomId == 0).List() as List<ItemData>;
            }
        }

        /// <summary>
        /// Get room items
        /// </summary>
        public static List<ItemData> GetRoomItems(int roomId)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                return session.QueryOver<ItemData>().Where(x => x.RoomId == roomId).List() as List<ItemData>;
            }
        }

        /// <summary>
        /// Get single item
        /// </summary>
        /// <returns>the item</returns>
        public static ItemData GetItem(int itemId)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                return session.QueryOver<ItemData>().Where(x => x.Id == itemId).Take(1).SingleOrDefault();
            }
        }

        /// <summary>
        /// Save item definition
        /// </summary>
        /// <param name="itemDefinition"></param>
        public static void SaveDefinition(ItemDefinitionData itemDefinition)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                using (var transaction = session.BeginTransaction())
                {
                    try
                    {
                        session.Update(itemDefinition);
                        transaction.Commit();
                    }
                    catch
                    {
                        transaction.Rollback();
                    }
                }
            }
        }

        /// <summary>
        /// Save item data
        /// </summary>
        public static void SaveItem(ItemData itemData)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                using (var transaction = session.BeginTransaction())
                {
                    try
                    {
                        session.Update(itemData);
                        transaction.Commit();
                    }
                    catch
                    {
                        transaction.Rollback();
                    }
                }
            }
        }


        /// <summary>
        /// Create items and refresh it with their filled in database ID's
        /// </summary>
        public static void CreateItems(List<ItemData> items)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                using (var transaction = session.BeginTransaction())
                {
                    try
                    {
                        foreach (var itemData in items)
                            session.Save(itemData);

                        transaction.Commit();

                        foreach (var itemData in items)
                            session.Refresh(itemData);
                    }
                    catch
                    {
                        transaction.Rollback();
                    }
                }
            }
        }

        /// <summary>
        /// Create item and refresh it with their filled in database ID's
        /// </summary>
        public static void CreateItem(ItemData item)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                using (var transaction = session.BeginTransaction())
                {
                    try
                    {
                        session.Save(item);
                        transaction.Commit();
                        session.Refresh(item);
                    }
                    catch
                    {
                        transaction.Rollback();
                    }
                }
            }
        }

        /// <summary>
        /// Delete item
        /// </summary>
        /// <param name="item"></param>
        public static void DeleteItem(ItemData item)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                session.Query<ItemData>().Where(x => x.Id == item.Id).Delete();
            }
        }
    }
}

File: euclid-storage/src/main/java/euclid/storage/database/access/ItemDao.java
package euclid.storage.database.access;

import euclid.storage.database.SessionFactoryBuilder;
import euclid.storage.database.data.ItemData;
import euclid.storage.database.data.ItemDefinitionData;
import org.hibernate.Session;
import org.hibernate.query.MutationQuery;
import org.hibernate.query.Query;

import java.util.List;

public class ItemDao {

    public static List<ItemDefinitionData> getDefinitions() {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            Query<ItemDefinitionData> q = session.createQuery("FROM ItemDefinitionData", ItemDefinitionData.class);
            return q.getResultList();
        }
    }

    public static List<ItemData> getUserItems(int userId) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            Query<ItemData> q = session.createQuery(
                    "FROM ItemData i WHERE i.ownerId = :userId AND i.roomId = 0", ItemData.class);
            q.setParameter("userId", userId);
            return q.getResultList();
        }
    }

    public static List<ItemData> getRoomItems(int roomId) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            Query<ItemData> q = session.createQuery(
                    "FROM ItemData i WHERE i.roomId = :roomId", ItemData.class);
            q.setParameter("roomId", roomId);
            return q.getResultList();
        }
    }

    public static ItemData getItem(int itemId) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            return session.find(ItemData.class, itemId);
        }
    }

    public static void saveDefinition(ItemDefinitionData itemDefinition) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            var transaction = session.beginTransaction();
            try {
                session.update(itemDefinition);
                transaction.commit();
            } catch (Exception e) {
                if (transaction.isActive()) {
                    transaction.rollback();
                }
            }
        }
    }

    public static void saveItem(ItemData itemData) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            var transaction = session.beginTransaction();
            try {
                session.update(itemData);
                transaction.commit();
            } catch (Exception e) {
                if (transaction.isActive()) {
                    transaction.rollback();
                }
            }
        }
    }

    public static void createItems(List<ItemData> items) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            var transaction = session.beginTransaction();
            try {
                for (ItemData item : items) {
                    session.persist(item);
                }
                transaction.commit();
                for (ItemData item : items) {
                    session.refresh(item);
                }
            } catch (Exception e) {
                if (transaction.isActive()) {
                    transaction.rollback();
                }
            }
        }
    }

    public static void createItem(ItemData item) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            var transaction = session.beginTransaction();
            try {
                session.persist(item);
                transaction.commit();
                session.refresh(item);
            } catch (Exception e) {
                if (transaction.isActive()) {
                    transaction.rollback();
                }
            }
        }
    }

    public static void deleteItem(ItemData item) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            var transaction = session.beginTransaction();
            try {
                MutationQuery query = session.createMutationQuery(
                        "DELETE FROM ItemData i WHERE i.id = :id");
                query.setParameter("id", item.getId());
                query.executeUpdate();
                transaction.commit();
            } catch (Exception e) {
                if (transaction.isActive()) {
                    transaction.rollback();
                }
            }
        }
    }
}
